Cannot generate expectations for aggregate_of covariates

The following test fails:

```
def test_make_df_from_expectations_with_minimum_of():
    study = StudyDefinition(
        population=patients.all(),
        a=patients.with_these_clinical_events(
            codelist(["A"], system="ctv3"),
            returning="numeric_value",
            find_first_match_in_period=True,
            return_expectations={
                "incidence": 0.2,
                "float": {"distribution": "normal", "mean": 60.0, "stddev": 60},
                "date": {"earliest": "2020-01-01", "latest": "2020-01-11"},
            },
        ),
        b=patients.with_these_clinical_events(
            codelist(["B"], system="ctv3"),
            returning="numeric_value",
            find_first_match_in_period=True,
            return_expectations={
                "incidence": 0.2,
                "float": {"distribution": "normal", "mean": 60.0, "stddev": 60},
                "date": {"earliest": "2020-01-01", "latest": "2020-01-11"},
            },
        ),
        m=patients.minimum_of(
            "a",
            "b",
            return_expectations={
                "incidence": 0.2,
                "float": {"distribution": "normal", "mean": 60.0, "stddev": 60},
                "date": {"earliest": "2020-01-01", "latest": "2020-01-11"},
            },
        ),
    )
    population_size = 10
    result = study.make_df_from_expectations(population_size)
    print(result)
```

It fails with:

```
$ pytest --tb=native -s tests/test_expectation_generators.py::test_make_df_from_expectations_with_minimum_of
======================================== test session starts ========================================
platform linux -- Python 3.8.2, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/inglesp/work/ebmdatalab/opensafely/cohort-extractor
plugins: freezegun-0.4.2
collected 1 item                                                                                    

tests/test_expectation_generators.py F

============================================= FAILURES ==============================================
__________________________ test_make_df_from_expectations_with_minimum_of ___________________________
Traceback (most recent call last):
  File "/home/inglesp/work/ebmdatalab/opensafely/cohort-extractor/tests/test_expectation_generators.py", line 824, in test_make_df_from_expectations_with_minimum_of
    study = StudyDefinition(
  File "/home/inglesp/work/ebmdatalab/opensafely/cohort-extractor/cohortextractor/study_definition.py", line 26, in __init__
    self._original_covariates = process_covariate_definitions(covariates)
  File "/home/inglesp/work/ebmdatalab/opensafely/cohort-extractor/cohortextractor/process_covariate_definitions.py", line 12, in process_covariate_definitions
    covariate_definitions = flatten_nested_covariates(covariate_definitions)
  File "/home/inglesp/work/ebmdatalab/opensafely/cohort-extractor/cohortextractor/process_covariate_definitions.py", line 40, in flatten_nested_covariates
    name, (query_type, query_args) = items.pop(0)
ValueError: too many values to unpack (expected 2)
====================================== short test summary info ======================================
FAILED tests/test_expectation_generators.py::test_make_df_from_expectations_with_minimum_of - Valu...
========================================= 1 failed in 0.47s =========================================

```

This is because `patients.minimum_of()` doesn't do anything special with `return_expectations`:

```
def minimum_of(*column_names, **extra_columns):
    aggregate_function = "MIN"
    column_names = column_names + tuple(extra_columns.keys())
    return "aggregate_of", locals()
```

I changed the signature to `minimum_of(*column_names, return_expectations=None, date_format=None, **extra_columns)`, and the test now passes.

But for date expressions it only generates years.  So

```
    study = StudyDefinition(
        population=patients.all(),
        a=patients.with_these_clinical_events(
            codelist(["A"], system="ctv3"),
            return_first_date_in_period=True,
            return_expectations={
                "incidence": 0.2,
                "date": {"earliest": "2020-02-01", "latest": "2021-02-01"},
            },
        ),
        b=patients.with_these_clinical_events(
            codelist(["B"], system="ctv3"),
            return_first_date_in_period=True,
            return_expectations={
                "incidence": 0.2,
                "date": {"earliest": "2020-02-01", "latest": "2021-02-01"},
            },
        ),
        m=patients.minimum_of(
            "a",
            "b",
            return_expectations={
                "incidence": 0.2,
                "date": {"earliest": "2020-02-01", "latest": "2021-02-01"},
            },
        ),
    )
```

generates eg

```
a     b     m
0   NaN   NaN   NaN
1  2021   NaN  2021
2   NaN   NaN   NaN
3   NaN  2020   NaN
4   NaN   NaN   NaN
5   NaN   NaN   NaN
6   NaN  2020   NaN
7   NaN   NaN  2020
8   NaN   NaN   NaN
9  2020   NaN   NaN
```

Adding a `date_format` parameter to `patients.minimum_of` doesn't work, and fails unsurprisingly with `TypeError: get_aggregate_expression() got an unexpected keyword argument 'date_format'`.

Fixing this might be quite involved, and we should maybe focus efforts on #359.

--- a/cohortextractor/study_definition.py
+++ b/cohortextractor/study_definition.py
@@ -175,8 +175,22 @@ class StudyDefinition:
         for name, (funcname, kwargs) in covariate_definitions.items():
             if name == "population" or kwargs.get("hidden"):
                 continue
-            args[name] = kwargs.copy()
+            kwargs = kwargs.copy()
+            kwargs["funcname"] = funcname
+            args[name] = kwargs
             column_type = kwargs["column_type"]
+
+            if funcname == "aggregate_of":
+                # We've already validated that the source columns for each
+                # aggregate expression are of the same type and, for dates,
+                # that they have the same date format. So we can just grab the
+                # first source column and use the date format from that.
+                other_column = kwargs["column_names"][0]
+                other_date_format = covariate_definitions[other_column][1].get(
+                    "date_format"
+                )
+                if other_date_format:
+                    kwargs["date_format"] = other_date_format
 
             # Awkward workaround: IMD is in fact an int, but it comes to us
             # rounded to nearest hundred which makes it act a bit more like a
@@ -230,6 +244,8 @@ class StudyDefinition:
         # matching on dependent columns
         for colname in self.pandas_csv_args["parse_dates"]:
             definition_args = self.pandas_csv_args["args"][colname]
+            if definition_args["funcname"] == "aggregate_of":
+                continue
             if "source" in definition_args:
                 source_args = self.pandas_csv_args["args"][definition_args["source"]]
                 definition_args["return_expectations"] = source_args[
@@ -257,6 +273,8 @@ class StudyDefinition:
         # its incidence calculated as a mask
         for colname, dtype in self.pandas_csv_args["dtype"].items():
             definition_args = self.pandas_csv_args["args"][colname]
+            if definition_args["funcname"] == "aggregate_of":
+                continue
             return_expectations = definition_args["return_expectations"] or {}
             if not self.default_expectations and not return_expectations:
                 raise ValueError(
@@ -293,6 +311,22 @@ class StudyDefinition:
                 raise ValueError(
                     f"Column definition {colname} does not return expected type {dtype}"
                 )
+
+        # Calculate the value of aggregated columns
+        for colname, definition_args in self.pandas_csv_args["args"].items():
+            if definition_args["funcname"] != "aggregate_of":
+                continue
+            fn = definition_args["aggregate_function"]
+            if fn == "MIN":
+                method_name = "min"
+            elif fn == "MAX":
+                method_name = "max"
+            else:
+                raise ValueError(f"Unsupported aggregate function '{fn}'")
+            columns = df[list(definition_args["column_names"])]
+            aggregate_method = getattr(columns, method_name)
+            df[colname] = aggregate_method(axis=1)
+
         # Finally, reduce date columns to the precision requested in
         # the definition
         for colname in self.pandas_csv_args["parse_dates"]:

--- a/tests/test_expectation_generators.py
+++ b/tests/test_expectation_generators.py
@@ -1,4 +1,4 @@
-from math import isclose
+from math import isclose, isnan
 
 import numpy as np
 import pandas as pd
@@ -816,3 +816,45 @@ def test_make_df_from_expectations_with_deregistration_date():
     dates = result.dereg_date.dropna()
     assert dates.max() <= "2018-02"
     assert dates.min() >= "1980-01"
+
+
+def test_make_df_from_expectations_with_aggregate_of():
+    study = StudyDefinition(
+        default_expectations={
+            "date": {"earliest": "1900-01-01", "latest": "today"},
+            "rate": "exponential_increase",
+            "incidence": 0.2,
+        },
+        population=patients.all(),
+        date_1=patients.with_these_clinical_events(
+            codelist(["X"], system="ctv3"),
+            returning="date",
+            date_format="YYYY-MM-DD",
+        ),
+        date_2=patients.with_these_clinical_events(
+            codelist(["X"], system="ctv3"),
+            returning="date",
+            date_format="YYYY-MM-DD",
+        ),
+        date_min=patients.minimum_of("date_1", "date_2"),
+        date_max=patients.maximum_of("date_1", "date_2"),
+    )
+    population_size = 10000
+    result = study.make_df_from_expectations(population_size)
+    for _, row in result.iterrows():
+        print(row)
+        dates = [d for d in [row["date_1"], row["date_2"]] if isinstance(d, str)]
+        if dates:
+            date_min = min(dates)
+            date_max = max(dates)
+        else:
+            date_min = float("nan")
+            date_max = float("nan")
+        assert_nan_equal(row["date_min"], date_min)
+        assert_nan_equal(row["date_max"], date_max)
+
+
+def assert_nan_equal(v1, v2):
+    assert (
+        not isinstance(v1, str) and not isinstance(v2, str) and isnan(v1) and isnan(v2)
+    ) or v1 == v2

